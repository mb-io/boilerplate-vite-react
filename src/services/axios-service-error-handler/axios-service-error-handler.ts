import { AxiosError } from 'axios'

import { Exception } from '@/exceptions/exception/exception'
import { Http403Exception } from '@/exceptions/http-403-exception/http-403-exception'
import { Http404Exception } from '@/exceptions/http-404-exception/http-404-exception'
import { Http500Exception } from '@/exceptions/http-500-exception/http-500-exception'
import { HttpException } from '@/exceptions/http-exception/http-exception'
import { HttpNoResponseException } from '@/exceptions/http-no-response-exception/http-no-response-exception'

export const axiosServiceErrorHandler = (error: AxiosError, message?: string) => {
  if (error.response) {
    // The request was made and the server responded with a status code
    // that falls out of the range of 2xx
    const status = error.response.status

    switch (true) {
      case status === 404:
        throw new Http404Exception(message, error)
      case status === 403:
        throw new Http403Exception(message, error)
      case status >= 500 && status <= 599:
        throw new Http500Exception(error.message, error)
      default:
        throw new HttpException(error.response.status, error.message, error)
    }
  } else if (error.request) {
    // The request was made but no response was received
    // `error.request` is an instance of XMLHttpRequest in the browser and an instance of
    // http.ClientRequest in node.js
    throw new HttpNoResponseException(error.message)
  } else {
    // Something happened in setting up the request that triggered an Error
    throw new Exception(error.message)
  }
}
