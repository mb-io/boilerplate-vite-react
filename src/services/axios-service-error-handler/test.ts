import { describe, it, expect } from 'vitest'
import { AxiosError, AxiosHeaders } from 'axios'
import { axiosServiceErrorHandler } from './axios-service-error-handler'
import { Http403Exception } from '@/exceptions/http-403-exception/http-403-exception'
import { Http404Exception } from '@/exceptions/http-404-exception/http-404-exception'
import { Http500Exception } from '@/exceptions/http-500-exception/http-500-exception'
import { HttpException } from '@/exceptions/http-exception/http-exception'
import { Exception } from '@/exceptions/exception/exception'

describe('axiosServiceErrorHandler', () => {
  it('should throw Http404Exception for 404 status', () => {
    const error = new AxiosError('Not Found', undefined, undefined, undefined, {
      status: 404,
      data: {},
      headers: {},
      statusText: 'Not Found',
      config: {
        headers: {} as AxiosHeaders
      }
    })
    expect(() => axiosServiceErrorHandler(error)).toThrow(Http404Exception)
    expect(() => axiosServiceErrorHandler(error)).toMatchSnapshot()
  })

  it('should throw Http403Exception for 403 status', () => {
    const error = new AxiosError('Not Authorized', undefined, undefined, undefined, {
      status: 403,
      data: {},
      headers: {},
      statusText: 'Not Authorized',
      config: {
        headers: {} as AxiosHeaders
      }
    })
    expect(() => axiosServiceErrorHandler(error)).toThrow(Http403Exception)
    expect(() => axiosServiceErrorHandler(error)).toMatchSnapshot()
  })

  it('should throw Http500Exception for 500 status', () => {
    const error = new AxiosError('Internal Server Error', undefined, undefined, undefined, {
      status: 500,
      data: {},
      headers: {},
      statusText: 'Internal Server Error',
      config: {
        headers: {} as AxiosHeaders
      }
    })
    expect(() => axiosServiceErrorHandler(error)).toThrow(Http500Exception)
    expect(() => axiosServiceErrorHandler(error)).toMatchSnapshot()
  })

  it('should throw HttpException for other status', () => {
    const error = new AxiosError('Bad Request', undefined, undefined, undefined, {
      status: 400,
      data: {},
      headers: {},
      statusText: 'Bad Request',
      config: {
        headers: {} as AxiosHeaders
      }
    })
    expect(() => axiosServiceErrorHandler(error)).toThrow(HttpException)
    expect(() => axiosServiceErrorHandler(error)).toMatchSnapshot()
  })

  it('should throw Exception for other errors', () => {
    const error = new AxiosError('Unknown Error')
    expect(() => axiosServiceErrorHandler(error)).toThrow(Exception)
    expect(() => axiosServiceErrorHandler(error)).toMatchSnapshot()
  })
})
