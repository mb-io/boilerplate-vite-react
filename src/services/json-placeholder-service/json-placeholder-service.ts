import axios, { AxiosRequestConfig, AxiosResponse } from 'axios'

import { axiosServiceErrorHandler } from '../axios-service-error-handler/axios-service-error-handler'

let requestCount = 0
let isThrottled = false

const resetThrottle = () => {
  requestCount = 0
  isThrottled = false
}

export interface JsonPlaceholderResponse {
  userId: number
  id: number
  title: string
  completed: boolean
}

export const jsonPlaceholderService = () => {
  const fetch = (options?: AxiosRequestConfig): Promise<JsonPlaceholderResponse> => {
    if (isThrottled) {
      throw new Error('Too many requests. Please wait a minute before trying again.')
    }

    if (requestCount >= 2) {
      isThrottled = true
      setTimeout(resetThrottle, 60000) // 1 minute
      throw new Error('Too many requests. Please wait a minute before trying again.')
    }

    requestCount += 1

    return axios
      .get(`/todos/1`, options)
      .then((response: AxiosResponse<JsonPlaceholderResponse>) => response.data)
      .catch((error) => {
        return axiosServiceErrorHandler(error, 'Failed to fetch placeholder data')
      })
  }

  return { fetch }
}
export type JsonPlaceholderService = AxiosResponse<JsonPlaceholderResponse>
