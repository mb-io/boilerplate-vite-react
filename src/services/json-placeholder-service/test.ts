import { describe, it, expect, vi, beforeEach, afterEach, Mock } from 'vitest'
import axios from 'axios'
import { jsonPlaceholderService } from './json-placeholder-service'
import { axiosServiceErrorHandler } from '../axios-service-error-handler/axios-service-error-handler'

vi.mock('axios')
vi.mock('../axios-service-error-handler/axios-service-error-handler')

describe('jsonPlaceholderService', () => {
  let service: ReturnType<typeof jsonPlaceholderService>

  beforeEach(() => {
    service = jsonPlaceholderService()
    vi.useFakeTimers()
  })

  afterEach(() => {
    vi.clearAllTimers()
    vi.resetAllMocks()
  })

  it('should fetch data successfully', async () => {
    const mockResponse = { data: { userId: 1, id: 1, title: 'Test Title', completed: false } }
    ;(axios.get as Mock).mockResolvedValue(mockResponse)

    const result = await service.fetch()

    expect(result).toEqual(mockResponse.data)
    expect(axios.get).toHaveBeenCalledWith('/todos/1', undefined)
    expect(axios.get).toHaveBeenCalledTimes(1)
  })

  it('should handle errors using axiosServiceErrorHandler', async () => {
    const mockError = new Error('Network Error')
    ;(axios.get as Mock).mockRejectedValue(mockError)
    ;(axiosServiceErrorHandler as unknown as Mock).mockImplementation((error: Error, message: string) => {
      throw new Error(message)
    })

    await expect(service.fetch()).rejects.toThrow('Failed to fetch placeholder data')
    expect(axiosServiceErrorHandler).toHaveBeenCalledWith(mockError, 'Failed to fetch placeholder data')
  })

  it('should throttle requests after 2 attempts', async () => {
    const mockResponse = { data: { userId: 1, id: 1, title: 'Test Title', completed: false } }
    ;(axios.get as Mock).mockResolvedValue(mockResponse)

    try {
      await service.fetch()
      await service.fetch()
    } catch (error) {
      expect(error).toEqual(new Error('Too many requests. Please wait a minute before trying again.'))
    }

    try {
      await expect(service.fetch()).rejects.toThrow('Too many requests. Please wait a minute before trying again.')
    } catch (error) {
      expect(error).toEqual(new Error('Too many requests. Please wait a minute before trying again.'))
    }

    // Ensure the throttle is reset after 1 minute
    vi.runAllTimers() // Fast-forward time by 1 minute

    try {
      await service.fetch()
    } catch (error) {
      expect(error).toEqual(new Error('Failed to fetch placeholder data'))
    }
    expect(axios.get).toHaveBeenCalledTimes(1)
  })

  it('should reset throttle after 1 minute', async () => {
    const mockResponse = { data: { userId: 1, id: 1, title: 'Test Title', completed: false } }
    ;(axios.get as Mock).mockResolvedValue(mockResponse)

    try {
      await service.fetch()
      await service.fetch()
    } catch (error) {
      expect(error).toEqual(new Error('Too many requests. Please wait a minute before trying again.'))
    }

    try {
      await expect(service.fetch()).rejects.toThrow('Too many requests. Please wait a minute before trying again.')
    } catch (error) {
      expect(error).toEqual(new Error('Too many requests. Please wait a minute before trying again.'))
    }

    vi.runAllTimers() // Fast-forward time by 1 minute

    const result = await service.fetch()
    expect(result).toEqual(mockResponse.data)
    expect(axios.get).toHaveBeenCalledTimes(2)
  })

  it('should match the snapshot of fetched data', async () => {
    const mockResponse = { data: { userId: 1, id: 1, title: 'Test Title', completed: false } }
    ;(axios.get as Mock).mockResolvedValue(mockResponse)

    const result = await service.fetch()

    expect(result).toMatchSnapshot()
  })
})
