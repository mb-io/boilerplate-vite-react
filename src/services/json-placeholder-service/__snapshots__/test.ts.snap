// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`jsonPlaceholderService > should match the snapshot of fetched data 1`] = `
{
  "completed": false,
  "id": 1,
  "title": "Test Title",
  "userId": 1,
}
`;
